1733 cleaning up theme resources

wdcossey opened the pull request:
- **What does the pull request do?**
  Fixing theme resources.
  - [x] Renamed ErrorBrushLight to ErrorLightBrush.
  - [x] ListBoxItem.xaml
  - [x] AutoCompleteBox.xaml
  - [x] CalendarButton.xaml
  - [x] CalendarDayButton.xaml
  - [x] CalendarItem.xaml
  - [x] DataValidationErrors.xaml
  - [x] DatePicker.xaml
  - [x] DropDownItem.xaml
  - [x] Expander.xaml
  - [x] ScrollBar.xaml
  - [x] ~~ScrollViewer.xaml~~ (*won't do*)
  - [x] Slider.xaml
  - [x] ~~TabStripItem.xaml~~ (*won't do*)
  - [x] TreeViewItem.xaml
  - [x] Added ListBox to ControlCatalog (Demo)
  - [x] Fix Slider
  - [x] Fix NumericUpDownPage.xaml (caused app crash)
  - [x] Added Typeface.Default.
  - [x] FormattedTextImpl to use Typeface.Default if typeface == null.

- **What is the current behavior?**
Not all default theme brushes are using Accents resources.

- **What is the updated/expected behavior with this PR?**
Use only Colors/Brushes defined in the resources.
A few minor bug fixes.

------

ListBox:
- Reworked the styling for the ListView.

![avalonia_listbox_sample](https://user-images.githubusercontent.com/1089478/44629414-ffd8c000-a946-11e8-8fba-5cbfb4082292.gif)

------

TreeView:
- Reworked the styling for the TreeView (looks like the ListView).
- Full row selection for TreeViewItem (like [VisualStudio](https://stackoverflow.com/questions/27398781/how-to-highlight-entire-row-in-wpf-treeview)).
- Added (readonly) **Depth** property to TreeViewItem (required for indentation).

![avalonia_treeview_sample](https://user-images.githubusercontent.com/1089478/44694369-ec237b80-aa64-11e8-9023-f48d36cce110.gif)

------

DropDown:
- List will inherit the style from **ListBox**

![avalonia_dropdown_sample](https://user-images.githubusercontent.com/1089478/44694531-e8dcbf80-aa65-11e8-9f38-922ffb7ca8cf.gif)

------

DatePicker:
- No more oversized **15** Path in the Button, uses the current day of the month (**not** the selected day).

![image](https://user-images.githubusercontent.com/1089478/44746408-327ae800-ab02-11e8-846c-e3e8d2c4af82.png)

------

Calendar:
- Today's date Foreground to use a lighter brush.

![image](https://user-images.githubusercontent.com/1089478/44747412-09a82200-ab05-11e8-829b-1b9399ab8c5f.png)

------

Fixes #1733 
Fixes #1775 
Fixes #1730

wdcossey committed:
Merge pull request #2 from AvaloniaUI/master

Merge

Commit:
Cleaning up Themes.
Renamed ErrorBrushLight to ErrorLightBrush.
ListBoxItem - Removed hardcoded Colors/Brushes, added additional sytling.
Removed ListBoxItem Style Selector from AutoCompleteBox (inherrited).
CalendarButton - Removed hardcoded Colors/Brushes.
DropDownItem - Removed hardcoded Colors/Brushes, added additional sytling.
ScrollBar - Removed hardcoded Colors/Brushes.
Added ListBox to ControlCatalog.
Slider bug fix (hopefully).

Commit:
Removed junk.

wdcossey committed:
Merge branch 'master' into 1733-cleaning-up-theme-resources

Commit:
Cleaned up resources for CalendarDayButton,CalendarItem,DataValidationErrors,DatePicker,Expander,Slider
Typeface - Added Typeface.Default.
FormattedTextImpl - Fallback to Typeface.Default for null typeface.

Commit:
Merge branch '1733-cleaning-up-theme-resources' of https://github.com/wdcossey/Avalonia into 1733-cleaning-up-theme-resources

Gillibald commented on src/Skia/Avalonia.Skia/FormattedTextImpl.cs:
```diff
@@ -30,6 +30,11 @@ public class FormattedTextImpl : IFormattedTextImpl
 
             SKTypeface skiaTypeface = TypefaceCache.Default;
 
+            if (typeface == null)
```
I think it is an error if FormattedTextImpl is created with a null TypeFace. If we keep this you should add this to Direct2D1 too.

wdcossey replied:
@Gillibald You are correct, **FormattedTextImpl** shouldn't be created with a null TypeFace.

I will remove this code and will add correct *ScreenPage.cs* as a cleaner workaround.

Commit:
ScreenPage.cs - Bug fix.
TreeViewItem - Added Depth property.
BaseLight.xaml - Added InvisibleSelectionColor/Brush.
DatePicker.xaml - Improvments.
TreeViewItem.xaml - Reworked styling.
FormattedTextImpl.cs - Reverted previous changes.

Commit:
Added missing Converter.

wdcossey committed:
Merge branch 'master' into 1733-cleaning-up-theme-resources

wdcossey committed:
Merge branch 'master' into 1733-cleaning-up-theme-resources

Gillibald commented on src/Avalonia.Controls/TreeViewItem.cs:
```diff
@@ -65,6 +75,34 @@ public bool IsSelected
+        /// </summary>
+        public int Depth
+        {
+            get { return this.GetDepth(this); }
```
Isn't it possible to do this in the ContainerGenerator? Caching it would also be nice.

wdcossey replied:
@Gillibald I have a look into this, need to do a little housekeeping on the PR in any case.

grokys replied:
Yes this could be calculated when added to the logical tree 👍

wdcossey replied:
That's exactly where it lives now, essentially "cached" too...

grokys commented on src/Avalonia.Controls/TreeViewItem.cs:
```diff
@@ -65,6 +75,34 @@ public bool IsSelected
+
+        private static TreeViewItem GetParent(TreeViewItem item)
+        {
+            var parent = item.InheritanceParent;
```
One question: why use `InheritanceParent` here rather than just `Parent`?

wdcossey replied:
Changed that earlier... 👍

Commit:
TreeViewItem updates.

Commit:
Merge branch '1733-cleaning-up-theme-resources' of https://github.com/wdcossey/Avalonia into 1733-cleaning-up-theme-resources

Commit:
Reverted some tab indentation.

wdcossey commented:
hmmmm, AppVeyor failed on something unrelated to this PR, see it has happened with a number of others too.

Commit:
Merge branch 'master' into 1733-cleaning-up-theme-resources

grokys commented on src/Avalonia.Controls/TreeViewItem.cs:
```diff
@@ -65,6 +73,15 @@ public bool IsSelected
             set { SetValue(IsSelectedProperty, value); }
         }
 
+        /// <summary>
```
Nit: should be just `Gets the depth of the item.` as the setter is private.

wdcossey replied:
Ah yes, that is correct, will fix that...

grokys commented on src/Avalonia.Themes.Default/Accents/BaseLight.xaml:
```diff
@@ -2,29 +2,59 @@
-    <SolidColorBrush x:Key="ThemeAccentBrush4">#33119EDA</SolidColorBrush>
-    <SolidColorBrush x:Key="ErrorBrush">Red</SolidColorBrush>
-    <SolidColorBrush x:Key="ErrorBrushLight">#10ff0000</SolidColorBrush>
+    <Color x:Key="InvisibleSelectionColor">#01000000</Color>
```
What is `InvisibleSelectionColor` needed for and why is its color `#01000000`? That seems like a strange value. Why not use `Transparent`?

wdcossey replied:
It's actually quite strange and something I picked up working a lot with WPF in the past... Transparent on (some) controls (see Border) don't track the cursor/pointer events, so the truck is to make the background near invisible to get around this.

I'll post a demo/gif later.

grokys replied:
Hmm, I don't see why that would be the case. If you can find a simple repro, please let me know!

wdcossey replied:
Lol, turns out this work around is no longer needed, it did however give me trouble earlier.

grokys commented on src/Avalonia.Themes.Default/TreeViewItem.xaml:
```diff
@@ -48,15 +59,27 @@
-  <Style Selector="TreeViewItem:selected /template/ ContentPresenter#PART_HeaderPresenter">
-    <Setter Property="Background" Value="#fff0f0f0"/>
+
+  <Style Selector="TreeViewItem[IsSelected=false] /template/ Border#SelectionBorder[IsPointerOver=true]">
```
It's best to use pseudoclasses where possible for performance reasons. Changing the selectors to be as below saves 400kb on the TreeView page in ControlCatalog:

```
  <Style Selector="TreeViewItem /template/ Border#SelectionBorder[IsPointerOver=true]">
    <Setter Property="Background" Value="{DynamicResource ThemeControlHighlightMidBrush}"/>
  </Style>

  <Style Selector="TreeViewItem:selected /template/ Border#SelectionBorder">
    <Setter Property="Background" Value="{DynamicResource ThemeAccentBrush4}"/>
  </Style>

  ...etc
```

wdcossey replied:
I can do something like the following?

```
  <Style Selector="TreeViewItem /template/ Border#SelectionBorder:pointerover">
    <Setter Property="Background" Value="{DynamicResource ThemeControlHighlightMidBrush}"/>
  </Style>

  <Style Selector="TreeViewItem:selected /template/ Border#SelectionBorder">
    <Setter Property="Background" Value="{DynamicResource ThemeAccentBrush4}"/>
  </Style>

  <Style Selector="TreeViewItem:selected /template/ Border#SelectionBorder:focus">
    <Setter Property="Background" Value="{DynamicResource ThemeAccentBrush3}"/>
  </Style>

  <Style Selector="TreeViewItem:selected /template/ Border#SelectionBorder:pointerover">
    <Setter Property="Background" Value="{DynamicResource ThemeAccentBrush3}"/>
  </Style>

  <Style Selector="TreeViewItem:selected /template/ Border#SelectionBorder:pointerover:focus">
    <Setter Property="Background" Value="{DynamicResource ThemeAccentBrush2}"/>
  </Style>
```

grokys reviewed (changes requested):
Looks good, just a few small comments/suggestions.

wdcossey commented on src/Avalonia.Styling/LogicalTree/LogicalExtensions.cs:
```diff
@@ -87,5 +87,20 @@ public static bool IsLogicalParentOf(this ILogical logical, ILogical target)
             return target.GetLogicalAncestors().Any(x => x == logical);
         }
+
+        public static int CalculateDistanceFromLogicalParent<T>(this ILogical logical, int @default = -1) where T : class
```
Will do

Commit:
TreeViewItem.cs - Renamed Depth to Level.
TreeViewItem.cs - Moved CalculateDistanceFromLogicalParent from LogicalExtensions.
TreeViewItem.xaml - Cleaned up selectors.
PropertyEqualsSelector.cs - Default null to string.Empty.
ContentControl.cs - Added valid & invalid pseudoclasses for ContentProperty.
BaseLight.xaml - Removed unused resources.
CheckBox.xaml - Hide ContentPresenter#PART_ContentPresenter when Content is null.

Commit:
Merge branch 'master' into 1733-cleaning-up-theme-resources

grokys committed:
Merge branch 'master' into 1733-cleaning-up-theme-resources
